### DISCUSSION TITLE
Comparison with Ava

### DISCUSSION THREAD

**Discussion by shellscape** [vitest-dev/vitest#3499] [Ideas]
I happened across https://github.com/vitest-dev/vitest/blob/main/docs/guide/comparisons.md today, and it struck me that there's no comparison with Ava. Ava is arguably far more popular than Uvu, but there's a comparison for Uvu. Would be highly useful for the folks writing Vitest to add a comparison with Ava.

**Comment by sheremet-va** [comment]
PR welcome!

**Reply by shellscape** [reply]
My dude. Please take a look at a Github profile before replying with that. When you reply with that to a prolific open source contributor and maintainer, it's just rude and it reflects poorly upon you. If I had the necessary knowledge and nuance about Vitest, I would have already done so.

 It also completely ignores the fact that I addressed the folks "writing Vitest" directly. Poor response, there.

**Reply by sheremet-va** [reply]
Sorry, I did not intend to offend you. I was busy, saw a notification, and quickly replied.

I will explain why I replied like that. I am one of the "folks writing Vitest". Most comparisons were written even before Vitest 0.1 so its concept was not implemented yet, but it was already formed.

Other comparisons were written by people who are familiar with the tool - Cypress by Jess (who is no longer on Cypress team), WebdriverIO section was written by the author of WebdriverIO, uvu comparison was written by trpc core team member.

If you don't see any comparisons it means no one was willing to write one - we even have an issue from 2021 to write a comparison for mocha which is still unresolved.

**Comment by turadg** [comment]
Here's some fodder from GPT-4o:


# Comparing Vitest and Ava Test Runners

Vitest and Ava are both modern JavaScript test runners with their own unique approaches, making them suitable for different use cases. Here's a detailed comparison:

---

## 1. Philosophy and Design
### **Vitest**
- **Philosophy:** Aims to be a fast, modern, and feature-rich test runner optimized for the Vite ecosystem but usable outside of it.
- **Design:**
  - Built-in support for mocking, snapshot testing, and TypeScript.
  - Deep integration with Vite, leveraging its lightning-fast HMR and bundling.
  - Provides a Jest-like API for easier transition from Jest.
  - Focuses on speed and simplicity, with minimal configuration required.

### **Ava**
- **Philosophy:** Focuses on simplicity, isolation, and performance, emphasizing concurrent test execution.
- **Design:**
  - Tests run in isolated environments to prevent shared state issues.
  - Concurrency is a core feature, designed for fast test execution.
  - Minimalist API with no global test functions—tests are explicitly imported.

---

## 2. Speed
### **Vitest**
- **Strengths:**
  - Built on Vite, benefiting from fast incremental builds and optimizations.
  - Efficient handling of ESM and dependency pre-bundling.
- **Use Case:** Ideal for Vite-based projects or modern ESM workflows.

### **Ava**
- **Strengths:**
  - Runs tests concurrently in isolated processes.
  - Excellent for test suites requiring heavy parallelism.
- **Trade-off:** Process spawning can make it slower for small/medium test suites.

---

## 3. Ecosystem and Compatibility
### **Vitest**
- Compatible with Vite and similar tools like SvelteKit or Nuxt.
- Jest-like APIs (`describe`, `it`, `expect`) make migration easy.
- Supports TypeScript out of the box.
- Built-in mocking and snapshot testing.

### **Ava**
- A standalone tool, not tied to a specific ecosystem.
- Explicit imports; no global test functions, promoting modular code.
- TypeScript support available but requires extra configuration.
- Relies on external libraries for mocking and snapshots.

---

## 4. Test Isolation
### **Vitest**
- Uses Vite’s module system for isolation and dependency handling.
- Suitable for most frontend projects but not as robust as Ava’s worker-based isolation.

### **Ava**
- Each test file runs in its own process or worker thread.
- Ensures complete isolation, preventing shared state issues.

---

## 5. Concurrency
### **Vitest**
- Supports concurrency but prioritizes speed through dependency pre-bundling.

### **Ava**
- Concurrency is a core feature, running tests in parallel by default.
- Ideal for CPU-intensive or I/O-heavy tests.

---

## 6. Community and Documentation
### **Vitest**
- Rapidly growing due to its alignment with Vite and modern tooling.
- Strong community support, especially for frontend developers.

### **Ava**
- Mature and well-documented but less active growth compared to Vitest.
- Appeals to developers who prioritize isolation and explicitness.

---

## 7. Use Cases
### **Vitest**
- Best for Vite-based, SvelteKit, or Nuxt projects.
- Ideal for frontend and full-stack applications needing fast feedback.
- Suitable for teams transitioning from Jest due to the similar API.

### **Ava**
- Great for backend-heavy projects requiring strong test isolation.
- Suitable for CI/CD pipelines where concurrency is a priority.
- Fits projects requiring explicit and modular test setups.

---

## Summary

| Feature                  | **Vitest**                          | **Ava**                           |
|--------------------------|--------------------------------------|-----------------------------------|
| **Speed**                | Extremely fast (Vite optimized)     | Fast (concurrency-based)         |
| **Test Isolation**       | Moderate                           | Strong (per-file processes)      |
| **Concurrency**          | Supported                          | Core feature                     |
| **API**                  | Jest-like                          | Minimalist, explicit             |
| **TypeScript Support**   | Built-in                           | Available (requires setup)       |
| **Mocking/Snapshots**    | Built-in                           | External libraries needed        |
| **Best For**             | Vite-based and frontend projects   | Isolated and backend projects    |

---

### **Conclusion**
- If you're using a **Vite-based setup** or prefer a Jest-like experience, **Vitest** is likely the better choice.
- If you need strong test isolation and concurrency for backend or complex systems, **Ava** might be more suitable.
